src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,555 @@
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\String;
+
+use Symfony\Component\String\Exception\ExceptionInterface;
+use Symfony\Component\String\Exception\InvalidArgumentException;
+
+/**
+ * Represents a string.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Hugo Hamon <hugohamon@neuf.fr>
+ *
+ * @throws ExceptionInterface
+ */
+abstract class AbstractString implements \JsonSerializable
+{
+    protected $string = '';
+    protected $ignoreCase = false;
+    protected $locale = null;
+
+    abstract public function __construct(string $string = '', string $locale = null);
+
+    public function toGrapheme(): GraphemeString
+    {
+        return new GraphemeString($this->string, $this->locale);
+    }
+
+    public function toUtf8(): Utf8String
+    {
+        return new Utf8String($this->string, $this->locale);
+    }
+
+    public function toBinary(): BinaryString
+    {
+        return new BinaryString($this->string, $this->locale);
+    }
+
+    /**
+     * @return static
+     */
+    public function ignoreCase(): self
+    {
+        $str = clone $this;
+        $str->ignoreCase = true;
+
+        return $str;
+    }
+
+    /**
+     * @return static
+     */
+    abstract public function after(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static
+     */
+    abstract public function afterLast(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static
+     */
+    abstract public function append(string ...$suffix): self;
+
+    /**
+     * @return static
+     */
+    abstract public function before(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static
+     */
+    abstract public function beforeLast(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static[]
+     */
+    abstract public function chunk(int $length = 1): array;
+
+    /**
+     * @return static
+     */
+    public function collapseWhitespace(): self
+    {
+        $str = clone $this;
+        $str->string = trim(preg_replace('/(?:\s{2,}+|[^\S ])/', ' ', $str->string));
+
+        return $str;
+    }
+
+    abstract public function equalsTo(string $string): bool;
+
+    abstract public function endsWith(string $suffix): bool;
+
+    /**
+     * @return static
+     */
+    public function ensureStart(string $prefix): self
+    {
+        return $this->startsWith($prefix) ? clone $this : $this->prepend($prefix);
+    }
+
+    /**
+     * @return static
+     */
+    public function ensureEnd(string $suffix): self
+    {
+        return $this->endsWith($suffix) ? clone $this : $this->append($suffix);
+    }
+
+    public function isEmpty(): bool
+    {
+        return '' === $this->string;
+    }
+
+    public function isUtf8(): bool
+    {
+        return '' === $this->string || preg_match('//u', $this->string);
```

Comment:
For the record, last time I benchmarked utf8 string detection algorithms, [this](https://github.com/rybakit/msgpack.php/blob/76e62c4ba34984cf575ccc04f999b7bdde45076c/src/Packer.php#L20-L29) regex was faster than `//u` (read this [thread](https://www.reddit.com/r/PHP/comments/54h70q/new_msgpackphp_v020_release_with_40_performance/d81u1i6/) for details), don't know if it makes sense to apply this micro-optimization in this library though.

Reply:
is this still the case nowadays? do you have some numbers?

Reply:
I've just checked it (lower is better):

```
utf8:

S regex            0.037
S //u              0.055
M regex            0.118
M //u              0.217
L regex            0.933
L //u              1.853

non-utf8:

S regex            0.031
S //u              0.024
M regex            0.075
M //u              0.106
L regex            0.479
L //u              0.920
```

Gist:  https://gist.github.com/rybakit/2c75152577fdcb9f4718d44e7123a539